src/java.base/share/classes/sun/reflect/annotation/AnnotationInvocationHandler.java

```diff
@@ -239,7 +249,11 @@ private static String toSourceString(Class<?> clazz) {
             arrayBrackets.append("[]");
         }
 
-        return finalComponent.getName() + arrayBrackets.toString() + ".class";
+        // Guard against NPE in toString generation if the class lacks
+        // a canonical name; not expected to happen
+        return Objects.toString(finalComponent.getCanonicalName(),
+                                "<no canonical name>") +
+            arrayBrackets.toString() + ".class";
```

Comment:
Since we're using the canonical name now (which takes the array brackets into account), can't the whole method be simplified down to the following?

```suggestion
        return Objects.toString(clazz.getCanonicalName(), "<no canonical name>") + ".class";
```

Reply:
The getCanonicalName is not specified to behave that way, should be a RFE I suppose, but appears to in practice; changed as suggested in subsequent push. Thanks.

Reply:
Filed https://bugs.openjdk.java.net/browse/JDK-8281671

Reply:
Thanks for filing that.